serviceAccountSecretRef is not always required

baelish opened the pull request:
**What this PR does / why we need it**:
Validation prevents you from creating an issuer/clusterIssuer with a clouddns solver without providing a serviceAccountSecretRef. The code however is able to use ambient
permissions.

**Which issue this PR fixes**
fixes #2249

**Special notes for your reviewer**:
This validation did not occur in v0.8, and definitely did in v0.10.1 and v0.11.0
I have tested this by downloading the current v0.11.0 yaml file editing it locally and applying the crds after. Then follow the rest of the instructions and was able to get both a normal and a wildcard certificate to be created and signed by letsencrypt using dns validation.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Allow clouddns resolvers to be validated correctly without `serviceAccountSecretRef` to allow ambient permissions to be used.
```

munnerz commented:
Thanks for the PR 😄 

This file is actually automatically generated by the `hack/update-crds.sh` script. Could you also add a `// +optional` tag to the field here: https://github.com/jetstack/cert-manager/blob/master/pkg/apis/acme/v1alpha2/types_issuer.go#L230 as well as marking it as `omitempty`, and converting it to be a pointer value?

It should look like this in the end:
```
// ACMEIssuerDNS01ProviderCloudDNS is a structure containing the DNS
// configuration for Google Cloud DNS
type ACMEIssuerDNS01ProviderCloudDNS struct {
	// +optional
	ServiceAccount *cmmeta.SecretKeySelector `json:"serviceAccountSecretRef,omitempty"`
	Project        string                   `json:"project"`
}
```

You'll probably also need to update a couple of places where this field is referenced to check for a nil value too 😄

/ok-to-test

munnerz commented:
/area api
/kind bug

baelish commented:
/retest

baelish commented:
> Thanks for the PR 
> 
> This file is actually automatically generated by the `hack/update-crds.sh` script. Could you also add a `// +optional` tag to the field here: https://github.com/jetstack/cert-manager/blob/master/pkg/apis/acme/v1alpha2/types_issuer.go#L230 as well as marking it as `omitempty`, and converting it to be a pointer value?
> 
> It should look like this in the end:
> 
> ```
> // ACMEIssuerDNS01ProviderCloudDNS is a structure containing the DNS
> // configuration for Google Cloud DNS
> type ACMEIssuerDNS01ProviderCloudDNS struct {
> 	// +optional
> 	ServiceAccount *cmmeta.SecretKeySelector `json:"serviceAccountSecretRef,omitempty"`
> 	Project        string                   `json:"project"`
> }
> ```
> 
> You'll probably also need to update a couple of places where this field is referenced to check for a nil value too 
> 
> /ok-to-test

Thanks for giving me a start for doing it properly. The crds file is generating the way I think it should but a test is still failing. I'll work on figuring out why the current tests don't pass when I've some time. It's slow going as I'm no go expert.

baelish commented:
@munnerz @JoshVanL With some excellent assistance from some of my colleagues I believe all is now well with this PR. Would you please check and progress this PR? Thanks.

JoshVanL commented on pkg/internal/apis/certmanager/validation/issuer.go:
```diff
@@ -284,10 +284,13 @@ func ValidateACMEChallengeSolverDNS01(p *cmacme.ACMEChallengeSolverDNS01, fldPat
+			// or serviceAccount.key is set, we validate the entire secret key
+			// selector
+			if p.CloudDNS.ServiceAccount != nil {
+				if p.CloudDNS.ServiceAccount.Name != "" || p.CloudDNS.ServiceAccount.Key != "" {
```
Why are we checking whether the strings are empty here? An empty string should be the same as a non-empty string for validation in this context. It should be enough to just check `ServiceAccount != nil` I should think. @munnerz 

The rest is looking good to me though :)

baelish replied:
Well the checks were there before when service account always required so figured they would still be desired in the scenario where a service account was provided. I suppose when someone goes through the trouble of setting a service account it could be considered enough reason to press on with the test rather then checking also whether the name/key is set first. I'll amend.

baelish commented:
@JoshVanL  I've made the change you requested.

JoshVanL commented:
Sorry @baelish, I have just spotted this https://github.com/jetstack/cert-manager/blob/313cd31c1d9b0286d2e1f6d2a89e0d94467b145a/pkg/issuer/acme/dns/dns.go#L236

Looks like this will cause a panic if the service account isn't set.

We need to check again for nil instead of empty string on the `Name`

baelish commented:
> Sorry @baelish, I have just spotted this
> 
> https://github.com/jetstack/cert-manager/blob/313cd31c1d9b0286d2e1f6d2a89e0d94467b145a/pkg/issuer/acme/dns/dns.go#L236
> 
> Looks like this will cause a panic if the service account isn't set.
> 
> We need to check again for nil instead of empty string on the `Name`

No worries @JoshVanL , looks like it. I've changed it and updated the commit.

JoshVanL commented:
Looking good to me, will leave it to @munnerz to lgtm
/approve

JoshVanL commented:
/unassign
/assign @munnerz

munnerz commented:
/lgtm

munnerz commented:
/cherrypick release-0.11
